### PR TITLE
As3 4.2

### PR DESCRIPTION
**Pull request by pascalf1309**
Added PlayerState APIs, UserItem service, ItemCatalog service and a few other APIs

### CODE REVIEW

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        /**
+         * 
+         *
+         * Service Name - customEntity
+         * Service Operation - GET_PAGE
+         *
+         * @param entityType
+         * @param rowsPerPage
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: missing 2 params, not a big deal though

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                "rowsPerPage":rowsPerPage,
+                "doCount": doCount
+            };
+
+            
+                data.searchJson  = searchJson;
+                data.sortJson  = sortJson;
+            
+
+            var serverCall:ServerCall = new ServerCall(ServiceName.customEntity, ServiceOperation.getPage, data, successCallback, errorCallback, cbObject);
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: can you confirm this is a small g on GetPage? Im pretty sure its a capital

Comment (ruth0048) [reply brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: meaning it wouldnt work if you made tests

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            Client.sendRequest(serverCall);
+        }
+
+                /**
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: SUMMARY >:(

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            var data:Object = {
+                "entityType": entityType,
+                "entityId":entityId,
+                "version": version
+            };
+
+            
+                data.dataJson  = dataJson;
+                data.acl  = acl;
+                data.timeToLive =timeToLive;
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: pretty sure this is not an object

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+            
+                data.fieldsJson  = fieldsJson;
+            
+
+            var serverCall:ServerCall = new ServerCall(ServiceName.CustomEntity, ServiceOperation.updateEntityFields, data, successCallback, errorCallback, cbObject);
+            Client.sendRequest(serverCall);
+        }
+    }
+}
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: ENTER!

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	{
+		private var _client:BrainCloudClient;
+		
+		public function BrainCloudItemCatalog(client:BrainCloudClient)
+		{
+			_client = client;
+		}
+        
+        /**
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: ...summary

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                "defId": defId
+            };	
+
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.ItemCatalog, ServiceOperation.GetCatalogItemDefinition, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+        
+        /**
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: summary

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+         * 
+         *
+         * @param context
+         * @param sortObject
+         * @param sortCriteria
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function getCatalogItemsPage(context:Object, searchCriteria:Object, sortCriteria:Object, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: Theres only one parameter for this call... Context

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		{			
+
+            data.context  = context;
+            data.searchCriteria = searchCriteria;
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: DESTROY THIS

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            data.sortCriteria  = sortCriteria;
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: DELETE ME PLZ :)

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            var data:Object = {
+                "context": context,
+                "pageOffset":pageOffset
+            };	
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.ItemCatalog, ServiceOperation.GetCatalogItemsPageOffset, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+	}
+}
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudItemCatalog.as]: ENTER!

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudPlayerState.as [hunk 12]
```diff
@@ -273,6 +273,88 @@ package com.bitheads.braincloud.services
...
+                "durationSecs":durationSecs
+            };	
+
+            data.details  = details;
+            
+			var serverCall:ServerCall = new ServerCall(ServiceName.PlayerState, ServiceOperation.SetUserStatus, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
 	}
 }
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudPlayerState.as]: ENTER!!!!!!!!

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+         * award/create item to user
+         *
+         * @param defId
+         * @param quantity
+         * @param includeDef
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function AwardUserItem(defId:String, quantity:int, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: K you gotta check this with an older service thats tested... but not sure if its capital I or lower case i... in another one of your calls you use big I.

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+         * drop item
+         *
+         * @param defId
+         * @param quantity
+         * @param includeDef
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function dropUserItem(defId:String, quantity:int, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: still not sure if int has small or capital

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+		{			
+
+            var data:Object = {
+                "defId": defId,
+                "quantity":quantity,
+                "includeDef":includeDef
+            };		
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItems, ServiceOperation.DropUserItems, data, successCallback, errorCallback, cbObject);
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: I see an S there... its DropUserItem no s :)

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+         * @param cbObject The user object sent to the callback
+         */
+        public function getUserItem(itemId:String, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
+		{				
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.GetUserItem, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+
+        		        /**
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: SUMMARY!?

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+         * @param context
+         * @param searchCriteria
+         * @param sortCriteria
+         * @param includeDef
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function getUserItemPage(context:Object, searchCriteria:Object, sortCriteria:Object, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: its items with an s, getUserItemsPage

Comment (ruth0048) [reply brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: Also... Only context and includeDef are the params to this call

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+            var data:Object = {
+                "includeDef":includeDef
+            };		
+
+            data.context  = context;
+            data.searchCriteria = searchCriteria;
+            data.sortCriteria  = sortCriteria;
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.GetUserItemPage, data, successCallback, errorCallback, cbObject);
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: GetUserItemsPage

File: brainCloudClient/src/com/bitheads/braincloud/ServiceOperation.as [hunk 19]
```diff
@@ -259,6 +261,26 @@ package com.bitheads.braincloud
...
         public static const UpdateContactEmail:String = "UPDATE_CONTACT_EMAIL";
+        public static const ClearUserStatus:String = "CLEAR_USER_STATUS";
+        public static const ExtendUserStatus:String = "EXTEND_USER_STATUS";
+        public static const SetUserStatus:String = "SET_USER_STATUS";
+        public static const GetUserStats:String = "GET_USER_STATUS";
+
+        public static const AwardUserItem:String = "AWARD_USER_ITEM";
+        public static const DropUserItem:String = "DROP_USER_ITEM";
+        public static const GetUserItem:String = "GET_USER_ITEM";
+        public static const GetUserItemPage:String = "GET_USER_ITEM_PAGE";
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/ServiceOperation.as]: GetUserItemsPage

File: brainCloudClient/src/com/bitheads/braincloud/ServiceOperation.as [hunk 20]
```diff
@@ -259,6 +261,26 @@ package com.bitheads.braincloud
...
+        public static const GetUserItemPageOffset:String = "GET_USER_ITEM_PAGE_BY_OFFSET";
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/ServiceOperation.as]: GetUserItemsPageOffset

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+         *
+         *
+         * @param context
+         * @param pageOffset
+         * @param includeDef
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function getUserItemPageOffset(context:String, pageOffset:int, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: getUserItemsPageOffset

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+		{			
+
+            var data:Object = {
+                "context":context,
+                "pageOffset":pageOffset,
+                "includeDef":includeDef
+            };
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.GetUserItemPageOffset, data, successCallback, errorCallback, cbObject);
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: GetUserItemsPageOffset

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 23]
```diff
@@ -0,0 +1,275 @@
...
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function purchaseUserItem(defId:String, quantity:int, shopId:Object, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
+		{		
+
+            var data:Object = {
+                "defId":defId,
+                "quantity":quantity,
+                "shopId":shopId,
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: this guys an Object, he needs to be handled differently

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 24]
```diff
@@ -0,0 +1,275 @@
...
+         * @param cbObject The user object sent to the callback
+         */
+        public function sellUserItem(itemId:String, version:int, quantity:int, shopId:Object, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
+		{	
+
+            var data:Object = {
+                "itemId":itemId,
+                "version":version,
+                "quantity":quantity,
+                "shopId":shopId,
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: treat me differently please :)

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+                "includeDef":includeDef
+            };	
+
+            data.newItemData  = newItemData;
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.UseUserItem, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+	}
+}
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: ENTERRRRRR

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+         * @param dataJson
+         * @param successCallback The success callback
+         * @param errorCallback The failure callback.
+         * @param cbObject The user object sent to the callback
+         */
+        public function createEntity(entityType:String, dataJson:Object, acl:Object, timeToLive:Object, isOwned:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
+        {
+            var data:Object = {
+                "entityType": eventName,
+                "timeToLive": timeToLive,
```
[the remark this answers is not in the document]
Comment (stevemrjones) [reply brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudCustomEntity.as]: Number ?

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+            data.context  = context;
+            data.searchCriteria = searchCriteria;
+            data.sortCriteria  = sortCriteria;
+
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.GetUserItemsPage, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+
+        		        /**
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: still no summary here

File: brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+         */
+        public function getUserItem(itemId:String, includeDef:Boolean, successCallback:Function = null, errorCallback:Function = null, cbObject:Object = null):void
+		{				
+			var serverCall:ServerCall = new ServerCall(ServiceName.UserItem, ServiceOperation.GetUserItem, data, successCallback, errorCallback, cbObject);
+			_client.sendRequest(serverCall);
+		}
+
+        		        /**
+         * 
+         *
```
Comment (ruth0048) [inline brainCloudClient/src/com/bitheads/braincloud/services/BrainCloudUserItems.as]: still no summary here, take a look at the call in the api explorer and copy and paste the description here

**Pull request merged**